20231214-listUsers-api.md

```diff
@@ -0,0 +1,572 @@
+    message ListUsersResponse {
+      repeated string users = 1; // e.g. document:1
+    }
+    ```
+    
+    - **StreamedListUsers**
+    
+    ```go
+    rpc StreamedListUsers(StreamedListUsersRequest) returns (stream StreamedListUsersResponse)
+    
+    message StreamedListUsersRequest {
+        string store_id = 1;
+        string authorization_model_id = 2;
+        string object = 3;
+        string relation = 4;
+        string user_object_type = 5;
+        string user_relation = 6;
+        repeated TupleKey contextual_tuples = 7;
+        google.protobuf.Struct context = 8;
+    }
+    
+    // StreamedListUsersResponse represents a single streaming user result 
+    // returned from the streaming endpoint.
+    message StreamedListUsersResponse {        
+      string user = 1; // e.g. document:1
+    }
+    ```
+    
+- New server configurations (flags/config/env):
+
+  - `--listUsers-max-results (uint32)` - limits the maximum size of the results returned by ListUsers (if 0 all results are returned up to the deadline, otherwise the response is limited to this size)
+
+  - `--listUsers-deadline (duration)` - the timeout deadline for serving ListUsers or StreamedListUsers requests (default 3s)
+
+  - `--max-concurrent-reads-for-list-objects (uint32)` - the maximum allowed number of concurrent datastore reads in a single ListUsers or StreamedListUsers query
+
+### Limits and Deadlines
+The `ListUsers` API is a [unary RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#unary-rpc) while the `StreamedListUsers` is a [server-streaming RPC](https://grpc.io/docs/what-is-grpc/core-concepts/#server-streaming-rpc). Because of these semantic differences we have different behaviors around limits and deadlines. 
+
+The server flag `--listUsers-deadline` (mentioned above) sets an overall upper limit on the amount of time (e.g. the deadline) for serving a `ListUsers` or `StreamedListUsers` request. It controls how long time can be spent expanding results (e.g. finding relationships in the graph). If there are very few results and the deadline is generous, then the request will respond earlier than the deadline period. However, if there are many results and returning all of them would take longer then the deadline, then only the subset that can be returned in the deadline period are returned.
+
+The server flag `--listUsers-max-results` (mentioned above) will limit the size of the set of results return from the unary `ListUsers` endpoint. It sets a hard upper limit on how many results can be returned. If the actual result set is less than this limit then the subset *should* be promptly returned so as to avoid keeping the client waiting longer than necessary. If the actual result set is at least as large as the max results limit but the results haven't been computed before the `--listUsers-deadline` period, then only the results which have been computed up to that deadline will be returned.
+
+### Error Handling
+### Typed Public Wildcard
```

Comment:
Can we iterate on this solution?

I assume returning `user:*` is cheap. We can build that first, and later add an option to expand it to all users if we see users demanding it. 


@netthier, are you using 'user:*'? If so, would it make sense to notify every user in the system when an object is changed?

Reply:
I do not use it and it does not make much sense in my specific use case.
I believe that just returning `user:*` would make more sense here though, as I'm not sure what exactly it should expand to?
OpenFGA doesn't necessarily contain all users that would pass that check :thinking:

Reply:
We've since put a lot of thought into the typed public wildcards. See the new dedicated section on it! TLDR: wildcards are seeked-out by default and returned in a separate part of the response body for special treatment.